File: scripts/common/coroutine_mgr.lua
-- Copyright (c) 2022 TungFai Fong <iam@tungfaifong.com>

CoroutineMgr = CoroutineMgr or BaseClass()

CORORESULT = {
	SUCCESS = 0,
	TIMEOUT = 1,
}

function CoroutineMgr:__init()
	if self._instance then
		return
	end

	self.TIMEOUT = 5000

	self._IDX = 0
	self._coros = {}
	self._timer_handler = INVALID_TIMER_ID
end

function CoroutineMgr:Start()
	self:_CheckTimeout()
end

function CoroutineMgr:Stop()
	timer.RemoveTimer(self._timer_handler)
	self._timer_handler = INVALID_TIMER_ID
	self._coros = nil
end

function CoroutineMgr:Instance()
	if self._instance == nil then
		self._instance = self:New()
	end
	return self._instance
end

function CoroutineMgr:Create(func, ...)
	local args = {...}
	self._IDX = self._IDX + 1
	self._coros[self._IDX] = {["timeout"] = engine.NowMs() + self.TIMEOUT, ["coro"] = coroutine.create(function()
		func(self._IDX, table.unpack(args))
	end)}
	coroutine.resume(self._coros[self._IDX].coro)
end

function CoroutineMgr:Resume(coro_id, result, data)
	if self._coros[coro_id] then
		if not coroutine.resume(self._coros[coro_id].coro, result, data) then
			self._coros[coro_id] = nil
		end
	end
end

function CoroutineMgr:_CheckTimeout()
	local now = engine.NowMs()
	for k, v in pairs(self._coros) do
		if now >= v.timeout then
			if not coroutine.resume(self._coros[k].coro, CORORESULT.TIMEOUT, nil) then
				self._coros[k] = nil
			end
		end
	end
	self._timer_handler = timer.CreateTimer(self.TIMEOUT, function() self:_CheckTimeout() end)
end

function CO_SPAWN(func, ...)
	CoroutineMgr:Instance():Create(func, ...)	
end

File: src/common/coroutine/coroutine_mgr.cpp
// Copyright (c) 2022 TungFai Fong <iam@tungfaifong.com>

#include "coroutine_mgr.h"

#include "usrv/interfaces/timer_interface.h"

bool CoroutineMgr::Start()
{
	_CheckTimeout();
	return true;
}

void CoroutineMgr::Stop()
{
	timer::RemoveTimer(_timer_handler);
	_timer_handler = INVALID_TIMER_ID;
}

void CoroutineMgr::Spawn(std::function<coroutine()> func)
{
	auto id = _coroutines.Insert(std::move(std::make_shared<coroutine>(func())));
	auto coro = _coroutines[id];
	coro->timeout = StdNow() + ms_t(TIMEOUT);
	coro->promise().coro = coro;
	coro->resume();
}

void CoroutineMgr::Resume(COROID coro_id, CORORESULT result, std::string && data)
{
	auto coro = _coroutines[coro_id];
	if(!coro)
	{
		return;
	}
	coro->result = result;
	coro->data = std::move(data);
	coro->resume();
	if(coro->done())
	{
		_coroutines.Erase(coro_id);
	}
}

void CoroutineMgr::_CheckTimeout()
{
	auto now = StdNow();
	for(auto iter = _coroutines.begin(); iter != _coroutines.end();)
	{
		auto coro = (*iter).second;
		if(now >= coro->timeout)
		{
			coro->result = CORORESULT::CR_TIMEOUT;
			coro->resume();
			iter = _coroutines.Erase(iter);
		}
		else
		{
			++iter;
		}
	}
	_timer_handler = timer::CreateTimer(TIMEOUT, std::bind(&CoroutineMgr::_CheckTimeout, shared_from_this()));
}
